### PR TITLE
Add Router method to execute custom logic before popstate events

### LINKED ISSUE #4008: useFileSystemPublicRoutes issues
Comment by @gcpantazis:

For what it's worth, it looks like the case mentioned by most of the reporters here (i.e., server renders a 404s then flashes to an unexpected client-rendered page) was fixed incidentally as of `5.0.1-canary.5` (and specifically by #3749).

There's a few other cases that we could consider:

### Explicit navigation:

1) Create a `Link` that corresponds to a file-system route, even though you disabled it via `useFileSystemPublicRoutes`.
2) Click on that `Link`.
3) See the page, even though `useFileSystemPublicRoutes: false` implies you shouldn't.

### `popstate`:

1) Go to a URL that corresponds to a file-system route, even though you disabled it via `useFileSystemPublicRoutes`. Get a 404 from SSR.
2) Click on a link that takes you something else via client-side navigation.
3) Hit "back" on the browser.
4) You'll see the file-system route render instead of a 404.

Of the two, I'd say the first is maybe not a huge concern, since as you've opted to replace Next's default routing you should be responsible for client-side navigation to bad URLs. The second I think is a bigger issue (and perhaps can be a separate issue, since as far as I can tell most of the reporters' problems will be resolved in `5.0.1`).

Test case here: https://github.com/gcpantazis/nextjs-playground/tree/useFileSystemPublicRoutes

### PR DESCRIPTION
**Pull request by gcpantazis**
Fixes #4008, 
`popstate` allows the Router to client-render pages that should be blocked by `useFileSystemPublicRoutes: false`. [See comment here](https://github.com/zeit/next.js/issues/2682#issuecomment-370664352).

This commit adds a pre-hook into `Router`'s response to `popstate`, which allows the downstream user to inject their own functionality / cancel Router's response if they'd like to handle it themselves. See README update for docs.

Example usage: https://github.com/gcpantazis/nextjs-playground/tree/uFSPR-solutionA 

cc/ @timneutkens

### CODE REVIEW

**Comment by arunoda** [comment]
@gcpantazis I'm okay to take this feature this in. But not as a solution to #2682 
@timneutkens mentioned that, we fixed ^^ issue. Did you check the `canary`. 
He's testing on this at the moment.

File: lib/router/router.js [hunk 1]
```diff
@@ -51,6 +52,12 @@ export default class Router {
   }
 
   async onPopState (e) {
+    // If the downstream application returns falsy, return.
+    // They will then be responsible for handling the event.
+    if (!this._beforePopState(e.state)) {
```
Comment (arunoda) [inline lib/router/router.js]: Could you move this after the following if condition?

Comment (gcpantazis) [reply lib/router/router.js]: Done

**Comment by gcpantazis** [comment]
Thanks @arunoda + @timneutkens.

I did check `canary` this morning, it still breaks for the case in #4008 (but as @timneutkens noted the rest of #2682 is fixed).

**Review by timneutkens — APPROVED** [review]

**Comment by timneutkens** [comment]
Going to merge this in, can't think of a different way to solve this, maybe we'll change it in the future.

**Pull request merged**